R/ProcessCSV.R

```diff
@@ -345,6 +452,246 @@ ProcessCSV <- function(user_csv, user_action, sample_storage_type, search_type =
+        user_data[[nm]] = c(NA)
+      }
+    }
+
+    user_data = dplyr::mutate(user_data, row_number = row_number())
+
+    stopifnot("Required database column names not implemented" = !is.null(required_names))
+
+    matched_columns <- match(required_names, names(user_data))
+    if (any(is.na(matched_columns))) {
+      errmsg <- paste("Required database column is missing:", paste(required_names[is.na(matched_columns)], collapse = ", "))
+      stop(errmsg)
+    }
+
+    ## broad sweep check if missing any data in required fields
+    rs <- rowSums(is.na(user_data[, requires_data])) > 0
+    rn <- user_data[rs, ] %>% pull(row_number)
+
+    cs <- colSums(is.na(user_data[, requires_data])) > 0
+    cols <- colnames(user_data[, requires_data])[cs]
+
+    if (length(cols) > 0) {
+      df <- user_data[rn, c("row_number", cols)]
+      err <- .maybe_add_err(err, df, "Rows found with missing data", dbmap)
+    }
+
+    ## detect `k` or `K` and convert to numeric
+    vb <- grepl('k|K', user_data$density)
+    user_data$density[vb] <- str_replace(user_data$density[vb], "k|K", "")
+    user_data$density <- as.integer(user_data$density)
+    user_data$density[vb] <- user_data$density[vb] * 1000 # convert to numeric
+
+
+    ## check date formatting
+    # allowed_date_formats = c("%Y-%m-%d")
+    # parsed_dates <- lubridate::parse_date_time(user_data$date, allowed_date_formats, quiet = TRUE, exact = TRUE)
+    # err <- .maybe_add_err(err, user_data[is.na(parsed_dates), c("row_number", "date")], "Unrecognized date format.", dbmap)
+
+    user_data.1 = user_data %>%
+      dplyr::mutate(
+        strain2 = strsplit(strain, ";"),
+        percentage2 = strsplit(percentage, ";")
+      ) %>%
+      tidyr::unnest(cols = c(strain2, percentage2))
+
+
+    con <- DBI::dbConnect(RSQLite::SQLite(), database)
+    copy_to(con, user_data.1)
+
+    ## Check that control batch is recorded in the database already
+
+    df = tbl(con, "user_data.1") %>%
+      left_join(tbl(con, "study"), by = c("batch"="short_code")) %>%
+      filter(is.na(id)) %>%
+      select(row_number, batch) %>%
+      collect()
+
+    err <- .maybe_add_err(err, df, "Batch is not yet recorded", dbmap)
+
+    ## check that all strains are found
+    df = tbl(con, "user_data.1") %>%
+      left_join(tbl(con, "strain"), by = c("strain2"="name")) %>%
+      filter(is.na(id)) %>%
+      select(row_number, strain2) %>%
+      distinct() %>%
+      dplyr::rename(strain = strain2) %>%
+      collect()
+
+    err <- .maybe_add_err(err, df, "Strain found that is not recorded in the database", dbmap)
+
+    ## percentage does not equal 100
+
+    df = tbl(con, "user_data.1") %>%
+      select(row_number, percentage, percentage2) %>%
+      group_by(row_number, percentage) %>%
+      dplyr::mutate(percentage2=as.double(percentage2)) %>%
+      dplyr::mutate(percentage2=ifelse(is.na(percentage2), 0, percentage2)) %>%
+      dplyr::mutate(equals_100 = as.logical(sum(percentage2, na.rm=TRUE) == 100)) %>%
```

Comment:
Should make this a "close" check, i.e. if the strains are 33/33/33 then it won't add up to 100.